Sources/StreamChat/Workers/Background/MessageSender.swift

```diff
@@ -195,31 +207,36 @@ private class MessageSendingQueue {
             // Sort the messages and send the oldest one
             // If this proves to be a bottleneck in the future, we might
             // switch to using a custom `OrderedSet`
-            guard let request = self?.requests.sorted(by: { $0.createdLocallyAt < $1.createdLocallyAt }).first else { return }
+            guard let sortedRequests = self?.requests.sorted(by: { $0.createdLocallyAt < $1.createdLocallyAt }) else { return }
+            guard let request = sortedRequests.first else { return }
 
             self?.messageRepository.sendMessage(with: request.messageId) { [weak self] result in
-                guard let self else { return }
-                self.removeRequestAndContinue(request)
-                if let error = result.error {
-                    switch error {
-                    case .messageDoesNotExist,
-                         .messageNotPendingSend,
-                         .messageDoesNotHaveValidChannel:
-                        let event = NewMessageErrorEvent(messageId: request.messageId, error: error)
-                        self.eventsNotificationCenter.process(event)
-                    case let .failedToSendMessage(error):
-                        let event = NewMessageErrorEvent(messageId: request.messageId, error: error)
-                        self.eventsNotificationCenter.process(event)
-                    }
+                // If we failed to send a message we must fail all the next messages as well for preserving the message order
+                if let error = result.error, case .failedToSendMessage = error {
+                    sortedRequests.forEach { self?.handleSendRequestResult(request: $0, result: result) }
+                } else {
+                    self?.handleSendRequestResult(request: request, result: result)
```

Comment:
I wonder if this could be configurable 🤔 Some customers might not want this behaviour. For some apps it could be more important if the delivery is done, and have as less errors as possible. 

The current approach has the con that if you send 10 messages, and the 3rd one fails, you will need to manually retry the other 7 messages :/

Reply:
I was also thinking about this again. Sounds like we should go for a full solution instead which is when failure is encountered because of internet connection, we fail all the messages in the queue. As soon as websocket connects, we automatically retry all the failed messages.